chain/lf3/ec.go

```diff
@@ -0,0 +1,173 @@
+
+import (
+	"context"
+	"sort"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-f3"
+	"github.com/filecoin-project/go-f3/gpbft"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/big"
+
+	"github.com/filecoin-project/lotus/chain/actors/builtin/power"
+	"github.com/filecoin-project/lotus/chain/stmgr"
+	"github.com/filecoin-project/lotus/chain/store"
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+type ecWrapper struct {
+	ChainStore   *store.ChainStore
+	StateManager *stmgr.StateManager
+	Manifest     f3.Manifest
+}
+
+type tsWrapper struct {
+	inner *types.TipSet
+}
+
+func (ts *tsWrapper) Key() gpbft.TipSetKey {
+	return ts.inner.Key().Bytes()
+}
+
+func (ts *tsWrapper) Beacon() []byte {
+	entries := ts.inner.Blocks()[0].BeaconEntries
+	if len(entries) == 0 {
+		return []byte{}
+	}
+	return entries[len(entries)-1].Data
+}
+
+func (ts *tsWrapper) Epoch() int64 {
+	return int64(ts.inner.Height())
+}
+
+func (ts *tsWrapper) Timestamp() time.Time {
+	return time.Unix(int64(ts.inner.Blocks()[0].Timestamp), 0)
+}
+
+func wrapTS(ts *types.TipSet) f3.TipSet {
+	if ts == nil {
+		return nil
+	}
+	return &tsWrapper{ts}
+}
+
+// GetTipsetByEpoch should return a tipset before the one requested if the requested
+// tipset does not exist due to null epochs
+func (ec *ecWrapper) GetTipsetByEpoch(ctx context.Context, epoch int64) (f3.TipSet, error) {
+	ts, err := ec.ChainStore.GetTipsetByHeight(ctx, abi.ChainEpoch(epoch), nil, true)
+	if err != nil {
+		return nil, xerrors.Errorf("getting tipset by height: %w", err)
+	}
+	return wrapTS(ts), nil
+}
+
+func (ec *ecWrapper) GetTipset(ctx context.Context, tsk gpbft.TipSetKey) (f3.TipSet, error) {
+	tskLotus, err := types.TipSetKeyFromBytes(tsk)
+	if err != nil {
+		return nil, xerrors.Errorf("decoding tsk: %w", err)
+	}
+
+	ts, err := ec.ChainStore.GetTipSetFromKey(ctx, tskLotus)
+	if err != nil {
+		return nil, xerrors.Errorf("getting tipset by key: %w", err)
+	}
+
+	return wrapTS(ts), nil
+}
+
+func (ec *ecWrapper) GetHead(_ context.Context) (f3.TipSet, error) {
+	return wrapTS(ec.ChainStore.GetHeaviestTipSet()), nil
+}
+
+func (ec *ecWrapper) GetParent(ctx context.Context, tsF3 f3.TipSet) (f3.TipSet, error) {
+	var ts *types.TipSet
+	if tsW, ok := tsF3.(*tsWrapper); ok {
+		ts = tsW.inner
+	} else {
+		tskLotus, err := types.TipSetKeyFromBytes(tsF3.Key())
+		if err != nil {
+			return nil, xerrors.Errorf("decoding tsk: %w", err)
+		}
+		ts, err = ec.ChainStore.GetTipSetFromKey(ctx, tskLotus)
+		if err != nil {
+			return nil, xerrors.Errorf("getting tipset by key for get parent: %w", err)
+		}
+	}
+	parentTs, err := ec.ChainStore.GetTipSetFromKey(ctx, ts.Parents())
+	if err != nil {
+		return nil, xerrors.Errorf("getting parent tipset: %w", err)
+	}
+	return wrapTS(parentTs), nil
+}
+
+func (ec *ecWrapper) GetPowerTable(ctx context.Context, tskF3 gpbft.TipSetKey) (gpbft.PowerEntries, error) {
```

Comment:
PowerEntries are serializable, while PowerTable is not.
Also for the majority of tipsets, we don't need the gpbft representation of the power table, we just need a hash of entires.

Reply:
> PowerEntries are serializable, while PowerTable is not.

Since `PowerTable` should be reproducible from its entries, that seems like a complexity to capture inside the `PowerTable` marshaller.

> for the majority of tipsets, we don't need the gpbft representation of the power table, we just need a hash of entires.

For those cases we should explicitly pass a powertable CID? 


Neither of the comments above are blockers. This just occurs to me as a debt.